### DISCUSSION TITLE
[OIDC] Access Token claims not imported using Identity Provider Attribute Importer Mappers. Keycloak Bug ?

### DISCUSSION THREAD

**Discussion by rm3l** [keycloak/keycloak#8462] [Ideas]
Hi there !

(I sent this to the [keycloak-dev mailing list](https://groups.google.com/u/1/g/keycloak-dev/c/l01fmZjA4Gg) few months ago, but got no answer so far. So I'm trying here.)

We have configured an external OIDC Identity Provider with no UserInfo endpoint, and are trying to map claims present in the Access Token issued to Keycloak User Attributes, using an Attribute Importer Mapper.

The tooltip message on the Mapper type states:

```
Import declared claim if it exists in ID, access token or the claim set returned by the user profile endpoint into the specified user property or attribute.
```

But Keycloak seems unable to find the claims in the Access Token, with the following DEBUG messages:

```
iam_1     | 16:52:24,551 DEBUG [org.keycloak.broker.oidc.mappers.AbstractJsonUserAttributeMapper] (default task-6) Going to process JsonNode path role-coh-irn71429 on data null
iam_1     | 16:52:24,552 DEBUG [org.keycloak.broker.oidc.mappers.AbstractJsonUserAttributeMapper] (default task-6) Going to process JsonNode path iat on data null
```

To understand the error, I attached a debugger to the Keycloak process, and I noticed the following:

[AbstractClaimMapper#getClaimValue(BrokeredIdentityContext,String)](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimMapper.java#L73-L95), which looks for the claim first in the access token, then in the ID token and then in the UserInfo (if set), is getting called, which is actually the expected behavior.
But I noticed that the search in the access token is performed with this line:
https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/mappers/AbstractClaimMapper.java#L74

```java
JsonWebToken token = (JsonWebToken)context.getContextData().get(KeycloakOIDCIdentityProvider.VALIDATED_ACCESS_TOKEN);
```

But `KeycloakOIDCIdentityProvider.VALIDATED_ACCESS_TOKEN` is put to the context in [KeycloakOIDCIdentityProvider#processAccessTokenResponse()](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/KeycloakOIDCIdentityProvider.java#L69), which overrides [OIDCIdentityProvider#processAccessTokenResponse](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L253-L256).

The implementation of [OIDCIdentityProvider#processAccessTokenResponse](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L253-L256) is actually empty. And, when debugging, I found that the object I was inspecting was an instance of `OIDCIdentityProvider`.

Maybe I missed something, but I'm wondering if it would not actually make more sense to move the implementation provided in [KeycloakOIDCIdentityProvider#processAccessTokenResponse](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/KeycloakOIDCIdentityProvider.java#L66-L70) to [OIDCIdentityProvider#processAccessTokenResponse](https://github.com/keycloak/keycloak/blob/15.0.2/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L253-L256) instead.
Otherwise, I don't see how Keycloak would be able to find the claims in the Access Token when the Identity Provider is an instance of `OIDCIdentityProvider` and the claim is not in the ID Token and no UserInfo endpoint has been defined.

We are using Keycloak 10.0.1, but I noticed the latest 15.0.2 and development (339224578edd944fc3426458ef374276dc646132) versions of Keycloak still behave the same way.

Any help will be appreciated.
I would also be happy to submit a PR for this, if this issue is confirmed.

EDIT : I tested the changes mentioned here in a fork, and this fixed the issue for us.

Thanks.

**Comment by galzetta** [comment]
Hi @rm3l .
I'm also having the same issue with Keycloak 12.0.4 and I opened a bug report related to this here: https://github.com/keycloak/keycloak/issues/8961

Could you provide a pull request/a diff for the changes required to fix this problem? Maybe in this way the fix might be resolved faster.

Thank you a lot for debugging this!

**Reply by rm3l** [reply]
Hi @galzetta,

Thanks for your feedback.
Before opening an issue, I wanted to discuss the changes proposed with Keycloak folks via this discussion and their mailing list.
But thanks for opening an issue. I will contribute the changes upstream when I have some time.

**Reply by rm3l** [reply]
I am trying to see how I could include one or more test cases.

Regarding your issue, even though we have similar logs, I am not sure the reproduction steps are similar. In my case, our customer purposely had their custom IDP claims present in the Access Token only.
But in your case, the IDP claims are in the ID Token, right ?
In this case, maybe that's a different issue then.

**Reply by rm3l** [reply]
@galzetta I was not able to add a failing test case for the issue you reported. So maybe I missed something in your reproduction steps, or the issues are slightly different.
So I finally ended up reporting a separate issue (#9004) and will contribute the changes for this issue instead. Hopefully, the issue you reported will get fixed as well.

**Reply by rm3l** [reply]
FYI, I contributed the fix in #9005

**Comment by Frank-LR** [comment]
Hi @rm3l 
I've got the same issue on keycloak version 17.0.1 with Okta/PKCE:
"_Advanced Claim to Role_" finds claims in the "id_token" but not in the "access_token"

In wich keycloak version was this issue solved?

**Reply by rm3l** [reply]
I just saw your message. I guess you got an answer from https://github.com/keycloak/keycloak/pull/21627#issuecomment-1729279107

**Comment by rm3l** [comment]
https://github.com/keycloak/keycloak/issues/9004 has been fixed. So closing this discussion.
